src/garage/tf/policies/categorical_cnn_policy.py

```diff
@@ -101,9 +102,9 @@ def __init__(self,
         self._initialize()
 
     def _initialize(self):
-        state_input = tf.compat.v1.placeholder(tf.float32,
+        state_input = tf.compat.v1.placeholder(tf.uint8,
                                                shape=(None, ) + self.obs_dim)
-
+        state_input = normalize_pixel_observation(self.env_spec, state_input)
```

Comment:
since the only valid input to this primitive is an image, you should probably verify it's an image here, rather than writing a conditional into the utility function

Reply:
Is there a foolproof way of checking whether observations are images, since they are simply tensors? I was going off what was done in `garage.misc.tensor_utils.normalize_pixel_batch` (which also just checks the shape of the Box). I can check that the box boundaries are 255 all around which means they are likely pixels, though that also does not guarantee it. The other option is to a pass an optional flag to the policy that indicates whether or not the observations are images.

Reply:
there isn't, and that's one of the difficult parts of this PR.

without adding an `Image` type to akro (which i think is a good idea), there isn't a whole lot which can be done in this PR specifically to guarantee these types are safe.

however, what we *can* do is decide whose job type safety is. since CNN policies can't infer image types, once data reaches them, it must be too late to do anything about it. so it can't be the responsibility of the CNN policies to enforce type safety. similarly, helper functions couldn't possibly do this without types, either (which is the issue I raise here).

if we raise ourselves on level in the call tree, we will find ourselves in the algorithm. that means it must be the algorithm's job to enforce type safety -- it can find out whether the observation is an image by interrogating the environment class, which it actually has access to.

to summarize, there are two good options here. which you use is up to you.

1) Add an Image type to [akro](https://github.com/rlworkgroup/akro) (a package we maintain), get me or @krzentner to cut akro release 0.0.7, bump garage's akro version to 0.0.7, then ensure `GarageEnv` converts the EnvSpec of image-based environments to use the new `Image` type rather than `Box`)
2) Add a flag to the algorithm alerting it that it should expect an image input, check the observation space (inside the algorithm) to ensure it's consistent with an image, refuse to use CNN-based primitives unless this flag is turned on.

(1) Is the long-term solution, but there's something to be said for doing (2) first since it would allow you to close the issue faster.

Reply:
Option 1 sounds cleaner to me. I'll create a PR in the akro repo for the new image type.